mongo/with_transactions_test.go

```diff
@@ -259,6 +259,99 @@ func TestConvenientTransactions(t *testing.T) {
 		assert.True(t, ok, "expected context for abortTransaction to contain ctxKey")
 		assert.Equal(t, "foobar", ctxValue, "expected value for ctxKey to be 'world', got %s", ctxValue)
 	})
+	t.Run("commitTransaction timeout allows abortTransaction", func(t *testing.T) {
+		coll := db.Collection("test")
+		// Explicitly create the collection on server because implicit collection creation is not allowed in
+		// transactions for server versions <= 4.2.
+		err := db.RunCommand(bgCtx, bson.D{{"create", coll.Name()}}).Err()
+		assert.Nil(t, err, "error creating collection on server: %v\n", err)
+
+		session, err := client.StartSession()
+		assert.Nil(t, err, "StartSession error: %v", err)
+
+		defer func() {
+			session.EndSession(bgCtx)
+			_ = coll.Drop(bgCtx)
+		}()
+
+		err = WithSession(bgCtx, session, func(sessionContext SessionContext) error {
+			// Start transaction
+			err = session.StartTransaction()
+			assert.Nil(t, err, "StartTransaction error: %v", err)
+
+			// Insert a document
+			_, err := coll.InsertOne(sessionContext, bson.D{{"val", 17}})
+			assert.Nil(t, err, "InsertOne error: %v", err)
+
+			// Set a timeout of 0 for commitTransaction
+			commitTimeoutCtx, commitCancel := context.WithTimeout(sessionContext, 0)
+			defer commitCancel()
+
+			// commitTransaction results in context.DeadlineExceeded
+			commitErr := session.CommitTransaction(commitTimeoutCtx)
+			assert.True(t, IsTimeout(commitErr),
+				"expected timeout error error; got %v", commitErr)
+
+			// abortTransaction without error
+			abortErr := session.AbortTransaction(context.Background())
+			assert.Nil(t, abortErr, "AbortTransaction error: %v", abortErr)
+
+			return nil
+		})
+	})
+	t.Run("commitTransaction timeout does not retry", func(t *testing.T) {
+		// Set high withTransactionTimeout so infinite CommitLoop will result in test timeout
+		withTransactionTimeout = 10 * time.Minute
```

Comment:
This new test is a little strange. We want to make sure that `CommitTransaction()` is run once within `WithTransaction()`'s CommitLoop, and that a single timeout error is received.

I set `withTransactionTimeout` to 10 minutes here, so the CommitLoop will continue until it reaches a network timeout error. I configure a fail point for `commitTransaction` that blocks connection for 100ms. I then set a timeout on the context for `WithTransaction()` of only 10ms. Together, these timeouts should cause a 100ms long `commitTransaction` that exceeds the deadline of the `WithTransaction()` context and returns the appropriate error earlier than 10 minutes. If the CommitLoop doesn't return within 10 minutes, the test would time out and fail.

There might be a much easier way of testing this, so let me know if I'm missing something obvious haha.

Reply:
Awesome! I think this is a good integration test to check this timeout behavior.

As a clarification to your comment, "network timeout error" is a `context.DeadlineExceeded` here.